tests/test_xpath1_parser.py

```diff
@@ -405,32 +405,87 @@ def test_node_set_functions(self):
         self.check_selector("name(tst:B1)", root, 'tst:B1', namespaces={'tst': "http://xpath.test/ns", '': ''})
 
     def test_string_functions(self):
+        root = self.etree.XML(u'<root><a id="a_id"><b>some content</b><c> space     space    \t .</c></a></root>')
         self.check_value("string(10.0)", '10.0')
+
         self.check_value("contains('XPath','XP')", True)
         self.check_value("contains('XP','XPath')", False)
         self.wrong_type("contains('XPath', 20)")
         self.wrong_syntax("contains('XPath', 'XP', 20)")
+        self.check_selector("a[contains(@id, '_i')]", root, [root[0]])
+        self.check_selector("a[contains(@id, '_b')]", root, [])
+        self.check_selector("//b[contains(., 'c')]", root, [root[0][0]])
+        self.check_selector("//b[contains(., ' -con')]", root, [])
+        self.check_selector("//none[contains(., ' -con')]", root, [])
+
         self.check_value("concat('alpha', 'beta', 'gamma')", 'alphabetagamma')
-        self.wrong_type("concat('alpha', 10, 'gamma')")
```

Comment:
concat() should convert all args to strings - https://www.w3.org/TR/xpath-functions/#func-concat

Reply:
Yes ... concat() is different, I'm also checking this.

Reply:
I've to consider compatibility with XPath 1.0: https://www.w3.org/TR/2010/REC-xpath-functions-20101214/#xpath1-compatibility